crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -374,19 +382,29 @@ unsafe impl<D: QueryData + 'static, F: QueryFilter + 'static> SystemParam for Qu
 pub(crate) fn init_query_param<D: QueryData + 'static, F: QueryFilter + 'static>(
     world: &mut World,
     system_meta: &mut SystemMeta,
-    state: &QueryState<D, F>,
-) {
+    state: Option<QueryState<D, F>>,
+) -> Entity {
+    let component_id = world.register_component::<QueryStateWrapper<D, F>>();
+    let state: QueryState<D, F> = state.unwrap_or_else(|| {
+        QueryState::new_with_access(world, &mut system_meta.archetype_component_access)
+    });
+
+    let mut component_access = state.component_access.clone();
+    // The Query need access to its own state. Just to be consistent, this will not
+    // cause a conflict since QueryStateWrapper is immutable
+    component_access.add_component_read(component_id);
```

Comment:
This now cause fn system(_query: Query<Entity>, _world: DeferredWorld) {}` to fail because `DeferredWorld` expects empty read access. We can remove this or modify `DeferredWorld` to accept read to immutable components. I prefer the later one

Reply:
I think we don't need this now that you're using an immutable component.  We need to satisfy the safety requirements for the call to `get::<QueryStateWrapper<D, F>>()`, and when it was an ordinary mutable component then we could only do that by registering access here.  But now that it's immutable, we can satisfy the requirements by noting that there cannot be any mutable access to the component.  

(This might also affect `fn system(query1: Query<Entity>, query2: Query<EntityMut>)`, since the `EntityMut` in the second query would conflict with the `QueryStateWrapper` access from the first query.)